File: my-folder/0347-top-k-frequent-elements/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  const countMap = new Map();
  for (const num of nums) {
    countMap.set(num, (countMap.get(num) ?? 0) + 1);
  }
  const maxCount = Math.max(...countMap.values());

  const buckets = Array.from({ length: maxCount + 1 }, () => []);
  for (const [num, count] of countMap) {
    buckets[count].push(num);
  }

  const ans = [];
  for (let i = maxCount; i >= 0 && ans.length < k; --i) {
    ans.push(...buckets[i]);
  }

  return ans.slice(0, k);
};

File: my-folder/0347-top-k-frequent-elements/solution.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequency = {}
        bucket = [[] for _ in range(len(nums) + 1)]
        for n in nums:
            frequency[n] = frequency.get(n, 0) + 1
        for n, c in frequency.items():
            bucket[c].append(n)

        res = []
        for i in range(len(bucket) - 1, 0, -1):
            for num in bucket[i]:
                res.append(num)
                if len(res) == k:
                    return res
        return res


